ignition-server/controllers/tokensecret_controller.go

```diff
@@ -109,44 +107,50 @@ func (r *TokenSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	log := ctrl.LoggerFrom(ctx)
 	log.Info("Reconciling")
 
+	// The controller maintains an internal cache of tokens to payloads and the
+	// secrets from which they are derived. Although cache entries have a TTL and
+	// should expire naturally after that time, we also actively purge entries
+	// from the cache if we can detect that a secret with associated tokens has
+	// disappeared. The deletion handling code is just to reconcile the internal
+	// token cache with reality in a more timely fashion than TTL alone.
+	secretDeleted := false
 	tokenSecret := &corev1.Secret{}
 	if err := r.Client.Get(ctx, req.NamespacedName, tokenSecret); err != nil {
 		if apierrors.IsNotFound(err) {
 			log.Info("Token Secret not found", "request", req.String())
-			return ctrl.Result{}, nil
+			secretDeleted = true
+		} else {
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, err
-	}
-
-	// If the tokenSecret is not being deleted then ensure the finalizer.
-	if tokenSecret.DeletionTimestamp.IsZero() {
-		if !controllerutil.ContainsFinalizer(tokenSecret, finalizer) {
-			controllerutil.AddFinalizer(tokenSecret, finalizer)
-			if err := r.Update(ctx, tokenSecret); err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
-			}
+	} else {
+		if !tokenSecret.DeletionTimestamp.IsZero() {
+			secretDeleted = true
 		}
 	}
 
-	// If the tokenSecret is being deleted then delete from cache and remove finalizer.
-	if !tokenSecret.DeletionTimestamp.IsZero() {
-		log.Info("Removing token from cache", "tokenSecret", client.ObjectKeyFromObject(tokenSecret))
-		r.PayloadStore.Delete(string(tokenSecret.Data[TokenSecretTokenKey]))
-		if controllerutil.ContainsFinalizer(tokenSecret, finalizer) {
-			controllerutil.RemoveFinalizer(tokenSecret, finalizer)
-			if err := r.Update(ctx, tokenSecret); err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+	if secretDeleted {
+		// If the secret was deleted, ensure that any corresponding tokens for that
+		// secret are purged from the cache and return as there's nothing to do.
+		for _, k := range r.PayloadStore.Keys() {
+			v, ok := r.PayloadStore.Get(k)
+			if ok && v.SecretName == req.Name {
+				log.Info("Deleting token for missing secret", "tokenSecret", req.Name)
+				r.PayloadStore.Delete(k)
 			}
 		}
 		return ctrl.Result{}, nil
 	}
 
-	// If the tokenSecret is older than ttl then delete it.
+	// If the tokenSecret is older than ttl then purge its token from the cache
+	// and delete the secret.
 	timeLived := time.Since(tokenSecret.CreationTimestamp.Time)
 	if timeLived >= ttl {
 		log.Info("Deleting expired token", "tokenSecret", client.ObjectKeyFromObject(tokenSecret))
+		r.PayloadStore.Delete(string(tokenSecret.Data[TokenSecretTokenKey]))
```

Comment:
Is this needed? It seems like the deletion of the secret would trigger another reconcile that would delete the payload from the cache.

Reply:
This is more of a nit and doesn't really hurt anything other than having two code paths where we remove from the cache instead of just one.  Not having this fix is hurting CI so not looking to delay on this.

Reply:
Probably unnecessary, might be cleaner to have just 1 place where the cache is cleaned up

Reply:
I think we commented at the same time about a single point of cache cleanup being a nicer design, agree it's probably not worth doing the tag dance again for right now